File: Chapter.04/DataParallelism.Part1.CSharp/Benchmark.cs
using BenchmarkDotNet.Attributes;

namespace DataParallelism.Part1.CSharp
{
    [MemoryDiagnoser]
    [RPlotExporter]
    [RankColumn]
    public class BenchmarkPrimeSum
    {
        public int len;

        [Params(1_000_000, 10_000_000, 100_000_000)]
        public int N;

        [GlobalSetup]
        public void Setup()
        {
            len = N;
        }

        [Benchmark]
        public void Sequential()
        {
            PrimeNumbers.PrimeSumSequential(len);
        }

        [Benchmark]
        public void Parallel()
        {
            PrimeNumbers.PrimeSumParallel(len);
        }

        [Benchmark]
        public void ParallelLINQ()
        {
            PrimeNumbers.PrimeSumParallelLINQ(len);
        }

        [Benchmark]
        public void ParallelThreadLocal()
        {
            PrimeNumbers.PrimeSumParallelThreadLocal(len);
        }
    }

    [MemoryDiagnoser]
    [RPlotExporter]
    [RankColumn]
    public class BenchmarkMandelbrot
    {
        [Params(1000, 2000, 3000)] public int N;

        public int size;

        [GlobalSetup]
        public void Setup()
        {
            size = N;
        }

        [Benchmark]
        public void Sequential()
        {
            Mandelbrot.SequentialMandelbrot(size);
        }

        [Benchmark]
        public void Parallel()
        {
            Mandelbrot.ParallelMandelbrot(size);
        }

        [Benchmark]
        public void ParallelLINQ()
        {
            Mandelbrot.ParallelMandelbrotOversaturation(size);
        }

        [Benchmark]
        public void ParallelThreadLocal()
        {
            Mandelbrot.ParallelStructMandelbrot(size);
        }
    }
}

File: Chapter.04/DataParallelism.Part1.FSharp/Benchmark.fs
namespace DataParallelism.Part1.FSharp

module Benchmark =

    open BenchmarkDotNet.Attributes

    [<MemoryDiagnoser>]
    [<RPlotExporter; RankColumn>]
    type BenchmarkPrimeSum() =

        [<Params(1_000_000, 10_000_000, 100_000_000)>]
        [<DefaultValue>]
        val mutable N : int
        
        [<DefaultValue>]
        val mutable len : int
        
        [<GlobalSetup>]
        member this.Setup() =
            this.len <- this.N
            
        [<Benchmark>]    
        member this.Sequential() = PrimeNumbers.sequentialSum this.len   

        [<Benchmark>]    
        member this.Parallel() = PrimeNumbers.parallelSum this.len
        
        [<Benchmark>]    
        member this.ParallelStruct() = PrimeNumbers.parallelLinqSum this.len
        
        
    [<MemoryDiagnoser>]
    [<RPlotExporter; RankColumn>]
    type BenchmarkMandelbrot() =

        [<Params(1000, 2000, 3000)>]
        [<DefaultValue>]
        val mutable N : int
        
        [<DefaultValue>]
        val mutable size : int
        
        [<GlobalSetup>]
        member this.Setup() =
            this.size <- this.N
            
        [<Benchmark>]    
        member this.Sequential() = Mandelbrot.sequentialMandelbrot this.size   

        [<Benchmark>]    
        member this.Parallel() = Mandelbrot.parallelMandelbrot this.size    
        
        [<Benchmark>]    
        member this.ParallelStruct() = Mandelbrot.parallelMandelbrotStruct this.size    